### ISSUE TITLE
IPython not setting exit status if script errors

### ISSUE THREAD

**Issue by jeremiahbuddha** [ipython/ipython#6912]
I'm using the subprocess module (Popen) to run a script using IPython, and it doesn't appear to be setting the system status correctly for the process.
```python
    #(script run_test.py)
    import subprocess as proc
    command = "ipython test.py"

    p = proc.Popen( command, shell=True, cwd=None, stdout=proc.PIPE,
                stderr=proc.STDOUT )
    ( stdOut, stdErr ) = p.communicate()

    print p.returncode
    if p.returncode:
        raise Exception( stdOut )
```

In this example, "test.py" is just a simple script that raises (just raise Exception()). When I run this script from the command line (Redhat Enterprise 5.11) I get:

    > python run_test.py
    0

Now, if I change my command in "run_test.py" to use python (instead of IPython), I get:

    (using command = "python test.py")
    > python run_test.py
    1 
    Traceback (most recent call last):
      File "run_test.py", line 12, in <module>
        raise Exception( stdOut )
    Exception: Traceback (most recent call last):
      File "test.py", line 2, in <module>
        raise Exception()
    Exception

Any tips on how to get ipython to set the system error code correctly?

**Comment by Fairly** [comment]
I've been digging into this issue for a while. I found that IPython cannot set exit status now because both it provides a %run magic in interactive mode and it is a command line tool to run Python scripts. Within the %run magic, IPython must catch all exceptions to avoid exiting IPython itself, which makes it always returns normally after running a script. It is mentioned in the function's annotation.
```python
def safe_execfile(self, fname, *where, **kw):
    """A safe version of the builtin execfile().

    This version will never throw an exception, but instead print
    helpful error messages to the screen.  This only works on pure
    Python files with the .py extension.

    Parameters
    ----------
    fname : string
        The name of the file to be executed.
    where : tuple
        One or two namespaces, passed to execfile() as (globals,locals).
        If only one is given, it is passed as both.
    exit_ignore : bool (False)
        If True, then silence SystemExit for non-zero status (it is always
        silenced for zero status, as it is so common).
    raise_exceptions : bool (False)
        If True raise exceptions everywhere. Meant for testing.
    shell_futures : bool (False)
        If True, the code will share future statements with the interactive
        shell. It will both be affected by previous __future__ imports, and
        any __future__ imports in the code will affect the shell. If False,
        __future__ imports are not shared in either direction.
    """
```
I wonder if there is way to separate these two situations, like adding a parameter to tell whether it's in the interactive mode, then the function can decide whether to throw an exception. Could anyone give me any advices?

**Comment by minrk** [comment]
There certainly is - the easiest way is probably to record that an exception was raised when executed through the entrypoint, and then set the exit code based on that.

**Comment by minrk** [comment]
closed by #8908
